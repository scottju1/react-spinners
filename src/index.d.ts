import * as React from 'react';

export namespace ReactSpinners {
	interface CommonProps {
		color?: string;
		loading?: boolean;
	}

	interface BarLoaderProps extends CommonProps {
		height?: number;
		heightunit?: string;
		width?: number;
		widthunit?: string;
	}

	interface BeatLoaderProps extends CommonProps {
		margin?: string;
		size?: number;
		sizeunit?: string;
	}

	interface BounceLoaderProps extends CommonProps {
		size?: number;
		sizeunit?: string;
	}

	interface CircleLoaderProps extends CommonProps {
		size?: number;
		sizeunit?: string;
	}

	interface ClipLoaderProps extends CommonProps {
		size?: number;
		sizeunit?: string;
	}

	interface ClimbingBoxLoaderProps extends CommonProps {
		size?: number;
		sizeunit?: string;
	}

	interface DotLoaderProps extends CommonProps {
		margin?: string;
		size?: number;
		sizeunit?: string;
	}

	interface FadeLoaderProps extends CommonProps {
		height?: number;
		heightunit?: string;
		width?: number;
		widthunit?: string;
		margin?: string;
		radius?: number;
	}

	interface GridLoaderProps extends CommonProps {
		size?: number;
		sizeunit?: string;
	}

	interface HashLoaderProps extends CommonProps {
		margin?: string;
		size?: number;
		sizeunit?: string;
	}

	interface MoonLoaderProps extends CommonProps {
		margin?: string;
		size?: number;
		sizeunit?: string;
	}

	interface PacmanLoaderProps extends CommonProps {
		margin?: string;
		size?: number;
		sizeunit?: string;
	}

	interface PropagateLoaderProps extends CommonProps {
		size?: number;
		sizeunit?: string;
	}

	interface PulseLoaderProps extends CommonProps {
		margin?: string;
		size?: number;
		sizeunit?: string;
	}

	interface RingLoaderProps extends CommonProps {
		margin?: string;
		size?: number;
		sizeunit?: string;
	}

	interface RiseLoaderProps extends CommonProps {
		margin?: string;
		size?: number;
		sizeunit?: string;
	}

	interface RotateLoaderProps extends CommonProps {
		margin?: string;
		size?: number;
		sizeunit?: string;
	}

	interface ScaleLoaderProps extends CommonProps {
		height?: number;
		heightunit?: string;
		width?: number;
		widthunit?: string;
		margin?: string;
		radius?: number;
		radiusunit?: string;
	}

	interface SyncLoaderProps extends CommonProps {
		margin?: string;
		size?: number;
		sizeunit?: string;
	}
}

export class BarLoader extends React.Component<ReactSpinners.BarLoaderProps> {}
export class BeatLoader extends React.Component<ReactSpinners.BeatLoaderProps> {}
export class BounceLoader extends React.Component<ReactSpinners.BounceLoaderProps> {}
export class CircleLoader extends React.Component<ReactSpinners.CircleLoaderProps> {}
export class ClipLoader extends React.Component<ReactSpinners.ClipLoaderProps> {}
export class ClimbingBoxLoader extends React.Component<ReactSpinners.ClimbingBoxLoaderProps> {}
export class DotLoader extends React.Component<ReactSpinners.DotLoaderProps> {}
export class FadeLoader extends React.Component<ReactSpinners.FadeLoaderProps> {}
export class GridLoader extends React.Component<ReactSpinners.GridLoaderProps> {}
export class HashLoader extends React.Component<ReactSpinners.HashLoaderProps> {}
export class MoonLoader extends React.Component<ReactSpinners.MoonLoaderProps> {}
export class PacmanLoader extends React.Component<ReactSpinners.PacmanLoaderProps> {}
export class PropagateLoader extends React.Component<ReactSpinners.PropagateLoaderProps> {}
export class PulseLoader extends React.Component<ReactSpinners.PulseLoaderProps> {}
export class RingLoader extends React.Component<ReactSpinners.RingLoaderProps> {}
export class RiseLoader extends React.Component<ReactSpinners.RiseLoaderProps> {}
export class RotateLoader extends React.Component<ReactSpinners.RotateLoaderProps> {}
export class ScaleLoader extends React.Component<ReactSpinners.ScaleLoaderProps> {}
export class SyncLoader extends React.Component<ReactSpinners.SyncLoaderProps> {}
