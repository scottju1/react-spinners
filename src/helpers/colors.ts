type RgbaFunction = (color: string, opacity: number) => string;

export const calculateRgba: RgbaFunction = (color: string, opacity: number): string => {
  if (color[0] === "#") {
    color = color.slice(1);
  }

  if (color.length === 3) {
    let res: string = "";
    color.split("").forEach((c: string) => {
      res += c;
      res += c;
    });
    color = res;
  }

  let rgbValues: string = color
    .match(/.{2}/g)!
    .map((hex: string) => parseInt(hex, 16))
    .join(", ");

  return `rgba(${rgbValues}, ${opacity})`;
};
