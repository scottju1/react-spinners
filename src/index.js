import BeatLoader from './spinners/BeatLoader';
import BounceLoader from './spinners/BounceLoader';
import ClipLoader from './spinners/ClipLoader';
import DotLoader from './spinners/DotLoader';
import FadeLoader from './spinners/FadeLoader';
import GridLoader from './spinners/GridLoader';
import MoonLoader from './spinners/MoonLoader';
import PacmanLoader from './spinners/PacmanLoader';
import PulseLoader from './spinners/PulseLoader';
import RingLoader from './spinners/RingLoader';
import RiseLoader from './spinners/RiseLoader';

module.exports = {
  BeatLoader,
  BounceLoader,
  ClipLoader,
  DotLoader,
  FadeLoader,
  GridLoader,
  MoonLoader,
  PacmanLoader,
  PulseLoader,
  RingLoader,
  RiseLoader
}