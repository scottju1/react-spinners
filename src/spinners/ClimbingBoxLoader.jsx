import React from 'react';
import PropTypes from 'prop-types';
import assign from 'domkit/appendVendorPrefix';
import insertKeyframesRule from 'domkit/insertKeyframesRule';

/**
 * @type {object}
 */
const keyframes = {
  
}

/**
 * @type {string}
 */
const animationName = insertKeyframesRule(keyframes);

class Loader extends React.Component {

  /**
   * @return {object} object with ball properties
   */
  getBallStyle() {
    return {
      
    }
  }

  /**
   * @param  {number} i element index
   * @return {object} object with animation properties
   */
  getAnimationStyle(i) {
    

    return {
      animation: animation,
      animationFillMode: animationFillMode
    }
  }

  /**
   * @param  {number} i element index
   * @return {object} object with style properties
   */
  getStyle(i) {
    return assign(
      
    );
  }

  /**
   * @param {boolean} loading Check if loading
   * @return {ReactComponent | null} Returns Loader or null
   */
  renderLoader(loading) {
    if (loading) {
      return (
        <div id={this.props.id} className={this.props.className}>
          
        </div>
      );
    }

    return null;
  }

  render() {
    return this.renderLoader(this.props.loading);
  }
}

/**
 * @type {object}
 */
Loader.propTypes = {
  
}

/**
 * @type {object}
 */
Loader.defaultProps = {
  
}

export default Loader;