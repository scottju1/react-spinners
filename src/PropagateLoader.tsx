/** @jsx jsx */
import * as React from "react";
import { keyframes, css, jsx, SerializedStyles } from "@emotion/core";
import { Keyframes } from "@emotion/serialize";

import { sizeDefaults, parseLengthAndUnit } from "./helpers";
import { LoaderSizeProps } from "./interfaces";

// 1.5 4.5 7.5
const distance: number[] = [1, 3, 5];

const propagate: Keyframes[] = [
  keyframes`
      25% {transform: translateX(-${distance[0]}rem) scale(0.75)}
      50% {transform: translateX(-${distance[1]}rem) scale(0.6)}
      75% {transform: translateX(-${distance[2]}rem) scale(0.5)}
      95% {transform: translateX(0rem) scale(1)}
    `,
  keyframes`
      25% {transform: translateX(-${distance[0]}rem) scale(0.75)}
      50% {transform: translateX(-${distance[1]}rem) scale(0.6)}
      75% {transform: translateX(-${distance[1]}rem) scale(0.6)}
      95% {transform: translateX(0rem) scale(1)}
    `,
  keyframes`
      25% {transform: translateX(-${distance[0]}rem) scale(0.75)}
      75% {transform: translateX(-${distance[0]}rem) scale(0.75)}
      95% {transform: translateX(0rem) scale(1)}
    `,
  keyframes`
      25% {transform: translateX(${distance[0]}rem) scale(0.75)}
      75% {transform: translateX(${distance[0]}rem) scale(0.75)}
      95% {transform: translateX(0rem) scale(1)}
    `,
  keyframes`
      25% {transform: translateX(${distance[0]}rem) scale(0.75)}
      50% {transform: translateX(${distance[1]}rem) scale(0.6)}
      75% {transform: translateX(${distance[1]}rem) scale(0.6)}
      95% {transform: translateX(0rem) scale(1)}
    `,
  keyframes`
      25% {transform: translateX(${distance[0]}rem) scale(0.75)}
      50% {transform: translateX(${distance[1]}rem) scale(0.6)}
      75% {transform: translateX(${distance[2]}rem) scale(0.5)}
      95% {transform: translateX(0rem) scale(1)}
    `
];

class Loader extends React.PureComponent<LoaderSizeProps> {
  public static defaultProps = sizeDefaults(15);

  public style = (i: number): SerializedStyles => {
    const { size, color } = this.props;
    const { value, unit } = parseLengthAndUnit(size || Loader.defaultProps.size);

    return css`
      position: absolute;
      font-size: ${`${value / 3}${unit}`};
      width: ${`${value}${unit}`};
      height: ${`${value}${unit}`};
      background: ${color};
      border-radius: 50%;
      animation: ${propagate[i]} 1.5s infinite;
      animation-fill-mode: forwards;
    `;
  };

  public wrapper = (): SerializedStyles => {
    return css`
      position: relative;
    `;
  };

  public render(): JSX.Element | null {
    const { loading, css } = this.props;

    return loading ? (
      <div css={[this.wrapper(), css]}>
        <div css={this.style(0)} />
        <div css={this.style(1)} />
        <div css={this.style(2)} />
        <div css={this.style(3)} />
        <div css={this.style(4)} />
        <div css={this.style(5)} />
      </div>
    ) : null;
  }
}

export default Loader;
