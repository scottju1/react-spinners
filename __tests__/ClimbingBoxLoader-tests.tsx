import * as React from "react";
import { mount, ReactWrapper } from "enzyme";
import { matchers } from "jest-emotion";
expect.extend(matchers);

import ClimbingBoxLoader from "../src/ClimbingBoxLoader";
import { LoaderSizeProps } from "../src/interfaces";
import { sizeDefaults } from "../src/helpers";

describe("ClimbingBoxLoader", () => {
  let loader: ReactWrapper;
  let props: LoaderSizeProps;

  it("should match snapshot", () => {
    loader = mount(<ClimbingBoxLoader />);
    expect(loader).toMatchSnapshot();
  });

  it("should contain default props if no props are passed", () => {
    props = loader.props();
    expect(props).toEqual(sizeDefaults(15));
  });

  it("parent div should contain styles created using default props", () => {
    expect(loader.find("div div")).toHaveStyleRule("font-size", "15px");
  });

  it("children div should contain styles created using default props", () => {
    expect(loader.find("div div div")).toHaveStyleRule("border", "0.25em solid #000000");
    expect(loader.find("div div div")).toHaveStyleRule("border-left", "0.25em solid #000000");
  });

  it("should render null if loading prop is set as false", () => {
    loader = mount(<ClimbingBoxLoader loading={false} />);
    expect(loader.isEmptyRender()).toBe(true);
  });

  it("renders the correct color based on prop", () => {
    loader = mount(<ClimbingBoxLoader color="#e2e2e2" />);
    expect(loader.find("div div div")).not.toHaveStyleRule("border", "0.25em solid #000000");
    expect(loader.find("div div div")).toHaveStyleRule("border", "025em solid #e2e2e2");
  });

  it("renders the correct size for the parent div based on props", () => {
    loader = mount(<ClimbingBoxLoader size={18} />);
    expect(loader).not.toHaveStyleRule("height", "15px");
    expect(loader).not.toHaveStyleRule("width", "15px");
    expect(loader).toHaveStyleRule("height", "18px");
    expect(loader).toHaveStyleRule("width", "18px");
  });

  it("renders the correct size for the child div based on props", () => {
    loader = mount(<ClimbingBoxLoader size={18} />);
    expect(loader.find("div div")).not.toHaveStyleRule("height", "15px");
    expect(loader.find("div div")).not.toHaveStyleRule("width", "15px");
    expect(loader.find("div div")).toHaveStyleRule("height", "18px");
    expect(loader.find("div div")).toHaveStyleRule("width", "18px");
  });

  it("renders the correct heightUnit based on props", () => {
    loader = mount(<ClimbingBoxLoader sizeUnit="%" />);
    expect(loader.find("div div")).not.toHaveStyleRule("height", "15px");
    expect(loader.find("div div")).not.toHaveStyleRule("width", "15px");
    expect(loader.find("div div")).toHaveStyleRule("height", "15%");
    expect(loader.find("div div")).toHaveStyleRule("width", "15%");
  });

  it("renders the css override based on props", () => {
    loader = mount(<ClimbingBoxLoader css={"position: absolute; overflow: scroll;"} />);
    expect(loader).not.toHaveStyleRule("position", "relative");
    expect(loader).toHaveStyleRule("position", "absolute");
    expect(loader).toHaveStyleRule("overflow", "scroll");
  });
});
