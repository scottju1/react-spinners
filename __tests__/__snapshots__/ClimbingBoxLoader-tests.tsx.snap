// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircleLoader should match snapshot 1`] = `
@keyframes animation-0 {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  50% {
    -webkit-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
    transform: rotate(180deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-0 {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  50% {
    -webkit-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
    transform: rotate(180deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-0 {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  50% {
    -webkit-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
    transform: rotate(180deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-0 {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  50% {
    -webkit-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
    transform: rotate(180deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-0 {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  50% {
    -webkit-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
    transform: rotate(180deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

.emotion-5 {
  position: relative;
  width: 50px;
  height: 50px;
}

.emotion-0 {
  position: absolute;
  height: 50px;
  width: 50px;
  border: 1px solid #000000;
  border-radius: 100%;
  -webkit-transition: 2s;
  transition: 2s;
  border-bottom: none;
  border-right: none;
  top: 0%;
  left: 0%;
  -webkit-animation-fill-mode: "";
  animation-fill-mode: "";
  -webkit-animation: animation-0 1s 0s infinite linear;
  animation: animation-0 1s 0s infinite linear;
}

.emotion-1 {
  position: absolute;
  height: 45px;
  width: 45px;
  border: 1px solid #000000;
  border-radius: 100%;
  -webkit-transition: 2s;
  transition: 2s;
  border-bottom: none;
  border-right: none;
  top: 1.75%;
  left: 0.875%;
  -webkit-animation-fill-mode: "";
  animation-fill-mode: "";
  -webkit-animation: animation-0 1s 0.2s infinite linear;
  animation: animation-0 1s 0.2s infinite linear;
}

.emotion-2 {
  position: absolute;
  height: 40px;
  width: 40px;
  border: 1px solid #000000;
  border-radius: 100%;
  -webkit-transition: 2s;
  transition: 2s;
  border-bottom: none;
  border-right: none;
  top: 3.5%;
  left: 1.75%;
  -webkit-animation-fill-mode: "";
  animation-fill-mode: "";
  -webkit-animation: animation-0 1s 0.4s infinite linear;
  animation: animation-0 1s 0.4s infinite linear;
}

.emotion-3 {
  position: absolute;
  height: 35px;
  width: 35px;
  border: 1px solid #000000;
  border-radius: 100%;
  -webkit-transition: 2s;
  transition: 2s;
  border-bottom: none;
  border-right: none;
  top: 5.249999999999999%;
  left: 2.6249999999999996%;
  -webkit-animation-fill-mode: "";
  animation-fill-mode: "";
  -webkit-animation: animation-0 1s 0.6000000000000001s infinite linear;
  animation: animation-0 1s 0.6000000000000001s infinite linear;
}

.emotion-4 {
  position: absolute;
  height: 30px;
  width: 30px;
  border: 1px solid #000000;
  border-radius: 100%;
  -webkit-transition: 2s;
  transition: 2s;
  border-bottom: none;
  border-right: none;
  top: 7%;
  left: 3.5%;
  -webkit-animation-fill-mode: "";
  animation-fill-mode: "";
  -webkit-animation: animation-0 1s 0.8s infinite linear;
  animation: animation-0 1s 0.8s infinite linear;
}

<onlyUpdateForKeys(Loader)
  color="#000000"
  css={Object {}}
  loading={true}
  size={50}
  sizeUnit="px"
>
  <Loader
    color="#000000"
    css={Object {}}
    loading={true}
    size={50}
    sizeUnit="px"
  >
    <div
      className="emotion-5"
    >
      <div
        className="emotion-0"
      />
      <div
        className="emotion-1"
      />
      <div
        className="emotion-2"
      />
      <div
        className="emotion-3"
      />
      <div
        className="emotion-4"
      />
    </div>
  </Loader>
</onlyUpdateForKeys(Loader)>
`;

exports[`ClimbingBoxLoader should match snapshot 1`] = `
@keyframes animation-0 {
  0% {
    -webkit-transform: translate(0,-1em) rotate(-45deg);
    -ms-transform: translate(0,-1em) rotate(-45deg);
    transform: translate(0,-1em) rotate(-45deg);
  }

  5% {
    -webkit-transform: translate(0,-1em) rotate(-50deg);
    -ms-transform: translate(0,-1em) rotate(-50deg);
    transform: translate(0,-1em) rotate(-50deg);
  }

  20% {
    -webkit-transform: translate(1em,-2em) rotate(47deg);
    -ms-transform: translate(1em,-2em) rotate(47deg);
    transform: translate(1em,-2em) rotate(47deg);
  }

  25% {
    -webkit-transform: translate(1em,-2em) rotate(45deg);
    -ms-transform: translate(1em,-2em) rotate(45deg);
    transform: translate(1em,-2em) rotate(45deg);
  }

  30% {
    -webkit-transform: translate(1em,-2em) rotate(40deg);
    -ms-transform: translate(1em,-2em) rotate(40deg);
    transform: translate(1em,-2em) rotate(40deg);
  }

  45% {
    -webkit-transform: translate(2em,-3em) rotate(137deg);
    -ms-transform: translate(2em,-3em) rotate(137deg);
    transform: translate(2em,-3em) rotate(137deg);
  }

  50% {
    -webkit-transform: translate(2em,-3em) rotate(135deg);
    -ms-transform: translate(2em,-3em) rotate(135deg);
    transform: translate(2em,-3em) rotate(135deg);
  }

  55% {
    -webkit-transform: translate(2em,-3em) rotate(130deg);
    -ms-transform: translate(2em,-3em) rotate(130deg);
    transform: translate(2em,-3em) rotate(130deg);
  }

  70% {
    -webkit-transform: translate(3em,-4em) rotate(217deg);
    -ms-transform: translate(3em,-4em) rotate(217deg);
    transform: translate(3em,-4em) rotate(217deg);
  }

  75% {
    -webkit-transform: translate(3em,-4em) rotate(220deg);
    -ms-transform: translate(3em,-4em) rotate(220deg);
    transform: translate(3em,-4em) rotate(220deg);
  }

  100% {
    -webkit-transform: translate(0,-1em) rotate(-225deg);
    -ms-transform: translate(0,-1em) rotate(-225deg);
    transform: translate(0,-1em) rotate(-225deg);
  }
}

.emotion-3 {
  position: relative;
  width: 7.1em;
  height: 7.1em;
}

.emotion-2 {
  position: absolute;
  top: 50%;
  left: 50%;
  margin-top: -2.7em;
  margin-left: -2.7em;
  width: 5.4em;
  height: 5.4em;
  font-size: 15px;
}

.emotion-0 {
  position: absolute;
  left: 0;
  bottom: -0.1em;
  height: 1em;
  width: 1em;
  background-color: transparent;
  border-radius: 15%;
  border: 0.25em solid #000000;
  -webkit-transform: translate(0,-1em) rotate(-45deg);
  -ms-transform: translate(0,-1em) rotate(-45deg);
  transform: translate(0,-1em) rotate(-45deg);
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
  -webkit-animation: animation-0 2.5s infinite cubic-bezier(0.79,0,0.47,0.97);
  animation: animation-0 2.5s infinite cubic-bezier(0.79,0,0.47,0.97);
}

.emotion-1 {
  position: absolute;
  width: 7.1em;
  height: 7.1em;
  top: 1.7em;
  left: 1.7em;
  border-left: 0.25em solid #000000;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

<onlyUpdateForKeys(Loader)
  color="#000000"
  css={Object {}}
  loading={true}
  size={15}
  sizeUnit="px"
>
  <Loader
    color="#000000"
    css={Object {}}
    loading={true}
    size={15}
    sizeUnit="px"
  >
    <div
      className="emotion-3"
    >
      <div
        className="emotion-2"
      >
        <div
          className="emotion-0"
        />
        <div
          className="emotion-1"
        />
      </div>
    </div>
  </Loader>
</onlyUpdateForKeys(Loader)>
`;
